OutOfSchool/OutOfSchool.Encryption/Services/EUSignOAuth2Service.cs

```diff
@@ -0,0 +1,480 @@
+
+                    string[] cmpServers = { ca.cmpAddress };
+                    string[] cmpServersPorts = { "80" };
+                    IEUSignCP.GetCertificatesByKeyInfo(
+                        keyInfo, cmpServers, cmpServersPorts, out certsCMP);
+
+                    int i = 0;
+                    while (true)
+                    {
+                        try
+                        {
+                            IEUSignCP.GetCertificateFromSignedData(
+                                i, certsCMP, out infoEx, out cert);
+                            if (infoEx.subjectType == IEUSignCP.EU_SUBJECT_TYPE.END_USER)
+                            {
+                                certs.Add(cert);
+                            }
+                        }
+                        catch (IEUSignCP.EUSignCPException ex)
+                        {
+                            if (ex.errorCode != IEUSignCP.EU_WARNING_END_OF_ENUM)
+                            {
+                                throw;
+                            }
+
+                            break;
+                        }
+
+                        i++;
+                    }
+                }
+            }
+
+            // Збереження сертифікатів ос. ключа
+            foreach (var cert in certs)
+            {
+                IEUSignCP.SaveCertificate(cert);
+            }
+
+            // Зчитування ос. ключа
+            if (fromFile)
+            {
+                IEUSignCP.CtxReadPrivateKeyBinary(
+                    context,
+                    pKey,
+                    password,
+                    out info,
+                    out pkContext);
+            }
+            else
+            {
+                IEUSignCP.CtxReadPrivateKey(
+                    context,
+                    keyMedia,
+                    out info,
+                    out pkContext);
+            }
+
+            try
+            {
+                // Отримання сертифікату для підпису
+                IEUSignCP.CtxGetOwnCertificate(
+                    pkContext,
+                    IEUSignCP.EU_CERT_KEY_TYPE_DSTU4145,
+                    IEUSignCP.EU_KEY_USAGE_DIGITAL_SIGNATURE,
+                    out infoEx,
+                    out pkSignCert);
+            }
+            catch (Exception)
+            {
+                throw new Exception("The private key does not support signing described in ДСТУ 4145");
+            }
+
+            try
+            {
+                // Отримання сертифікату для направленого шифрування
+                IEUSignCP.CtxGetOwnCertificate(
+                    pkContext,
+                    IEUSignCP.EU_CERT_KEY_TYPE_DSTU4145,
+                    IEUSignCP.EU_KEY_USAGE_KEY_AGREEMENT,
+                    out infoEx,
+                    out pkEnvelopCert);
+            }
+            catch (Exception)
+            {
+                throw new Exception("The private key does not support PKI encryption described in ДСТУ 4145");
```

Comment:
We've lost the original exception here, could we log it with the error ? 
If it is taken as is from the example - it might be the risk to log due to sensitive data exposure...

Reply:
@h4wk13 according to their source code the exception text is generated by `C` code. No indication if it is sensitive or not.